ethcore/src/private_transactions/mod.rs

```diff
@@ -0,0 +1,907 @@
+		})
+	}
+}
+
+impl Encodable for SignedPrivateTransaction {
+	fn rlp_append(&self, s: &mut RlpStream) {
+		s.begin_list(4);
+		s.append(&self.private_transaction_hash);
+		s.append(&self.v);
+		s.append(&self.r);
+		s.append(&self.s);
+	}
+}
+
+impl SignedPrivateTransaction {
+	/// Construct a signed private transaction message
+	pub fn new(private_transaction_hash: H256, sig: Signature, chain_id: Option<u64>) -> Self {
+		SignedPrivateTransaction {
+			private_transaction_hash: private_transaction_hash,
+			r: sig.r().into(),
+			s: sig.s().into(),
+			v: sig.v() as u64 + if let Some(n) = chain_id { 35 + n * 2 } else { 27 },
+		}
+	}
+
+	/// 0 if `v` would have been 27 under "Electrum" notation, 1 if 28 or 4 if invalid.
+	pub fn standard_v(&self) -> u8 {
+		match self.v {
+			v if v == 27 => 0 as u8,
+			v if v == 28 => 1 as u8,
+			v if v > 36 => ((v - 1) % 2) as u8,
+			 _ => 4
+		}
+	}
+
+	/// Construct a signature object from the sig.
+	pub fn signature(&self) -> Signature {
+		Signature::from_rsv(&self.r.into(), &self.s.into(), self.standard_v())
+	}
+
+	/// Get the hash of of the original transaction.
+	pub fn private_transaction_hash(&self) -> H256 {
+		self.private_transaction_hash
+	}
+}
+
+struct TransactionDetailsProvider<'a> {
+	client: &'a MiningBlockChainClient,
+}
+
+impl<'a> TransactionDetailsProvider<'a> {
+	pub fn new(client: &'a MiningBlockChainClient) -> Self {
+		TransactionDetailsProvider {
+			client: client,
+		}
+	}
+}
+
+impl<'a> TransactionQueueDetailsProvider for TransactionDetailsProvider<'a> {
+	fn fetch_account(&self, address: &Address) -> AccountDetails {
+		AccountDetails {
+			nonce: self.client.latest_nonce(address),
+			balance: self.client.latest_balance(address),
+		}
+	}
+
+	fn estimate_gas_required(&self, tx: &SignedTransaction) -> U256 {
+		tx.gas_required(&self.client.latest_schedule()).into()
+	}
+
+	fn is_service_transaction_acceptable(&self, _tx: &SignedTransaction) -> Result<bool, String> {
+		Ok(false)
+	}
+}
+
+/// Configurtion for private transaction provider
+#[derive(Default, PartialEq, Debug)]
+pub struct ProviderConfig {
+	/// Accounts that can be used for validation
+	pub validator_accounts: Vec<Address>,
+	/// Account used for signing public transactions created from private transactions
+	pub signer_account: Option<Address>,
+	/// Passwords used to unlock accounts
+	pub passwords: Vec<String>,
+	/// Account used for signing requests to key server
+	pub key_server_account: Option<Address>,
+	/// URL to key server
+	pub key_server_url: Option<String>,
+	/// Key server's threshold
+	pub key_server_threshold: u32,
+}
+
+#[derive(Debug)]
+/// Private transaction execution receipt.
+pub struct Receipt {
+	/// Private transaction hash.
+	pub hash: H256,
+	/// Created contract address if any.
+	pub contract_address: Option<Address>,
+	/// Execution status.
+	pub status_code: u8,
+}
+
+/// Manager of private transactions
+pub struct Provider {
+	encryptor: RwLock<Arc<Encryptor>>,
```

Comment:
Since every part of this struct requires synchronization, why not wrap the entire struct into `Mutex/RwLock` instead?